korma/src/commonMain/kotlin/com/soywiz/korma/geom/Rectangle.kt

```diff
@@ -306,6 +306,19 @@ inline fun IRectangleInt.anchor(ax: Number, ay: Number): IPointInt = anchor(ax.t
 
 val IRectangleInt.center get() = anchor(0.5, 0.5)
 
+val IRectangleInt.topLeft get() = PointInt(left, top)
+val IRectangleInt.topRight get() = PointInt(right, top)
+val IRectangleInt.bottomLeft get() = PointInt(left, bottom)
+val IRectangleInt.bottomRight get() = PointInt(right, bottom)
+
+val IRectangleInt.x2: Int get() = x + width - 1
+val IRectangleInt.y2: Int get() = y + height - 1
+
+val IRectangleInt.topLeft2 get() = PointInt(x, y)
+val IRectangleInt.topRight2 get() = PointInt(x2, y)
+val IRectangleInt.bottomLeft2 get() = PointInt(x, y2)
+val IRectangleInt.bottomRight2 get() = PointInt(x2, y2)
```

Comment:
Can you provide samples/use-cases where the inner points are used? When I required to find out if a point was inside just used `<` instead of `<=`.
In the case they are needed could we use longer names like `rightInclusive` or `innerRight` or something like that?

Reply:
> Can you provide samples/use-cases where the inner points are used?

There are plenty examples. Let me demonstrate two.

The first one is when I want to shrink a line into a rectangle:
```kotlin
fun ClipHLine(pnt1: PointInt, x2: Mutable<Int>, rect: IRectangleInt): Boolean {
    if (pnt1.y !in rect.y..rect.y2) {
        return false
    }
    pnt1.x = pnt1.x.clamp(rect.x, rect.x2)
    x2.element = x2.element.clamp(rect.x, rect.x2)
    return pnt1.x != x2.element
}
```

The second one is when I want to draw the border of a rectangle:
```kotlin
fun FrameRect(rc: IRectangleInt, clr: IDibPixel, a: UByte = 255u) {
    val drect = RectangleInt(rc)
    if (!ClipRect(drect, GetSize().asRectangle())) {
        return
    }
    HLine(rc.topLeft2, rc.x2, clr, a)
    HLine(rc.bottomLeft2, rc.x2, clr, a)
    VLine(rc.topLeft2, rc.y2, clr, a)
    VLine(rc.topRight2, rc.y2, clr, a)
}
```

> When I required to find out if a point was inside just used `<` instead of `<=`.

Yep, I can use `rect.y until rect.bottom` instead of `rect.y..rect.y2` but it's just longer...

> In the case they are needed could we use longer names like `rightInclusive` or `innerRight` or something like that?

Again, it's just longer. All in all, it's for you to decide.

Reply:
Regarding to `until`, well actually I preferred the `...` operator from Haxe that worked as until, since I use much more `until` than just `..` and in other cases. That's one of the things from Kotlin that I'm not very happy with (having to use until all the way).
The first reason is that C, Java and Kotlin arrays are zero based. So typically for an array of 10, you do: `for (int n = 0; n < 10; n++)` 10 elements, from 0 to 9. In BASIC for example you would do 1 to 10, but it is not the case here.

For me `x2` would be an alias for `right` ( `x + width` ), and the same for `y2` and `bottom`. The 2 part is not intuitive, unless there is a 1 too.
Not sure what `HLine` and `VLine` do, and why they couldn't work by not including the last one in the iteration or comparisons. 

The `ClipHLine` would work the same like this:
```
fun ClipHLine(pnt1: PointInt, x2: Mutable<Int>, rect: IRectangleInt): Boolean {
    if (pnt1.y !in rect.y until rect.y2) {
        return false
    }
    pnt1.x = pnt1.x.clamp(rect.left, rect.right)
    x2.element = x2.element.clamp(rect.left, rect.right)
    return pnt1.x != x2.element
}
```

Not sure if clamping exactly would be required for the `x`. But instead of providing x2, maybe one could create `clampOpenRight(a, b)` or doing stuff in another way.

In general by using an exclusive end value, I don't remember having to provide -1 or +1 frequently.
But since we have different opinions here. Let's request further points of view.

Reply:
I have requested a few reviewers. As far as I know at least @TobseF and @avdim have experience in videogame development, maybe they have an opinion in this area to untie this?

Reply:
Yep, let's ask for other people :) Actually, I'm not insisting on including these `**2` properties – I've already placed them in sources of my project. Of course, it will be handier for me to have these properties right in the repo, but again, it's for your team to decide.

> Not sure what `HLine` and `VLine` do, and why they couldn't work by not including the last one in the iteration or comparisons.

They are quite the same so let's talk about `HLine` – it draws horizontal line with a width of 1 px:
```kotlin
fun HLine(pos: IPointInt, x2: Int, clr: IDibPixel, a: UByte)
```

The line's `y = pos.y` and `x = minOf(pos.x,x2)..maxOf(pos.x,x2)`. So seems like I need a closed range here.

> `clampOpenRight(a, b)`

Oh, it's just too long...

Reply:
Lets keep this a couple of days in the case someone else wants to join the conversation.

From my side I have a strong opinion on this:

All the libraries I have tried myself, define right as x + width. So I think thats the expected naming and result.

The **2 functions even if they are shorter don't provide information about their return value in their name. And they are not common enough (as far as Im concerned) for people to recognize them.

As a counter example Bit-wise operators for example << >> & | ~ are pretty common on lots of languages for people to recognize, so I would add them to Kotlin.

In this case I would prefer to add -1 to a few places to be understandable, than to add x2.

From my side I can consider adding long function names that are descriptive enough to tell what the property is returning or what the function is doing if other people think they are handy.

Then you can add shortcuts in your code like x2 or y2 if you are familiar and comfortable with them. In any case you can keep them in your code

Reply:
Another thing we can do is to create a source-code bundle so you and other people wanting these extensions can use them and even publish them to:

https://awesome.korge.org/

that way you can reuse that functionality in several projects without having deploy a full artifact/recompile on each version or having to duplicate code. And people finding them confusing can opt-out by not including the bundle in their projects